docs/CppInteroperability/SwiftToCppInteroperabilityVisionDocument.md

```diff
@@ -0,0 +1,137 @@
+
+## Overview of Swift's interoperability support
+
+Swift uses two very different approaches for interoperability with C and Objective-C. For "forward" interoperability, Swift embeds a copy of the Clang compiler, which it uses to directly load C and Objective-C Clang modules and translate declarations into a native Swift representation. For "reverse" interoperability, Swift does not want to require Clang to embed Swift, and so Swift generates an Objective-C header that can be used to call into Swift APIs that are exposed in that header. Because Objective-C is restricted in what kinds of types it can work with, and because it requires code to be emitted in a special way that doesn't match the regular Swift ABI, methods and types must be marked as being exposed to Objective-C with the `@objc` attribute.
+
+Swift's support for C++ interoperability is modeled after its support for C and Objective-C interoperability. For "forward" interoperability, the embedded Clang compiler is used to directly load C++ Clang modules and translate declarations into a native Swift representation. The "forward" interoperability [vision document](https://github.com/apple/swift/pull/60501/files) provides more details about this model. For "reverse" interoperability, Swift generates a header that uses C++ language constructs to represent Swift APIs that are exposed by the Swift module. Because C++ is much more expressive and flexible than Objective-C, the generated header is able to provide representation for native Swift functions, methods, initializers, accessors and types. This allows C++ programmers to call into Swift APIs using the familiar C++ function and member function call syntax.
+
+### C++ interoperability is opt-out not opt-in
+
+The user model employed by Swift-to-C++ interoperability is different than the user model employed by the Objective-C interoperability. C++ can use Swift's native calling convention when calling Swift functions, and C++ does not have restrictions on what Swift types it can work with. This means that Swift APIs do not require additional annotations when they're bridged to C++. This fact highlights the key difference between C++ and Objective-C interoperability; Swift library author is not required to change their code when their APIs are exposed to C++, whereas they are required to opt-in into interoperability using an `@objc` annotation when exposing APIs to Objective-C. Swift library authors will still be able to opt-out from exposing certain APIs to C++ when they desire to do so.
```

Comment:
Curious what the use case for the "opt-out" looks like. Are we concerned that there will be Swift constructs that we do expose to C++ as a matter of language policy but (due to their implementation on the Swift side) can't actually be used properly from C++?

Reply:
I omitted "opt-out" from the vision document now. I don't have a concrete example for "opt-out", but I can think of potential clients hiding function/method overloads that don't map nicely to C++ because of complicated overload signatures involving argument labels.

Reply:
Perhaps certain types/functions  in the standard library will need to be opted-out as well.